mayastor/src/subsys/mbus/registration.rs

```diff
@@ -0,0 +1,203 @@
+/// them.
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display(
+        "Failed to connect to the MessageBus server {}: {:?}",
+        server,
+        cause
+    ))]
+    ConnectFailed {
+        cause: std::io::Error,
+        server: String,
+    },
+    #[snafu(display(
+        "Cannot issue requests if message bus hasn't been started"
+    ))]
+    NotStarted {},
+    #[snafu(display("Failed to queue register request: {:?}", cause))]
+    QueueRegister { cause: std::io::Error },
+    #[snafu(display("Failed to queue deregister request: {:?}", cause))]
+    QueueDeregister { cause: std::io::Error },
+}
+
+/// Register message payload
+#[derive(Serialize, Deserialize, Debug)]
+struct RegisterArgs {
+    id: String,
+    #[serde(rename = "grpcEndpoint")]
+    grpc_endpoint: String,
+}
+
+/// Deregister message payload
+#[derive(Serialize, Deserialize, Debug)]
+struct DeregisterArgs {
+    id: String,
+}
+
+#[derive(Clone)]
+struct Configuration {
+    /// Name of the node that mayastor is running on
+    node: String,
+    /// gRPC endpoint of the server provided by mayastor
+    grpc_endpoint: String,
+    /// heartbeat interval (how often the register message is sent)
+    hb_interval: Duration,
+}
+
+#[derive(Clone)]
+pub struct Registration {
+    /// Configuration of the registration
+    config: Configuration,
+    /// Receive channel for messages and termination
+    rcv_chan: smol::channel::Receiver<()>,
+    /// Termination channel
+    fini_chan: smol::channel::Sender<()>,
+}
+
+static MESSAGE_BUS_REG: OnceCell<Registration> = OnceCell::new();
+impl Registration {
+    /// initialise the global registration instance
+    pub(super) fn init(node: &str, grpc_endpoint: &str) {
+        MESSAGE_BUS_REG.get_or_init(|| Registration::new(node, grpc_endpoint));
+    }
+
+    /// terminate and re-register
+    pub(super) fn fini(&self) {
+        self.fini_chan.close();
+    }
+
+    pub(super) fn get() -> &'static Registration {
+        MESSAGE_BUS_REG.get().unwrap()
+    }
+
+    /// runner responsible for registering and
+    /// de-registering the mayastor instance on shutdown
+    pub async fn run() -> Result<(), ()> {
+        if let Some(registration) = MESSAGE_BUS_REG.get() {
+            registration.clone().run_loop().await;
+        }
+        Ok(())
+    }
+
+    fn new(node: &str, grpc_endpoint: &str) -> Registration {
+        let (msg_sender, msg_receiver) = smol::channel::unbounded::<()>();
+        let config = Configuration {
+            node: node.to_owned(),
+            grpc_endpoint: grpc_endpoint.to_owned(),
+            hb_interval: match env::var("MAYASTOR_HB_INTERVAL")
+                .map(|v| v.parse::<u64>())
+            {
+                Ok(Ok(num)) => Duration::from_secs(num),
+                _ => HB_INTERVAL,
+            },
+        };
+        Self {
+            config,
+            rcv_chan: msg_receiver,
+            fini_chan: msg_sender,
+        }
+    }
+}
+
+impl Registration {
```

Comment:
why 2 impl blocks?

Reply:
no particular reason really, I'll fix that.